### ISSUE TITLE
Getting Resource UID Returns Invalid (-1) at Runtime on Exported Projects

### ISSUE THREAD

**Issue by TheoJao** [godotengine/godot#75617]
### Godot version

4.0.1

### System information

Windows 11

### Issue description

I'm building a character customization system in game, and I use custom resources to that contain information like the textures and thumbnails of parts of the face and stuff. When the character creator is loaded, I create the thumbnails and connect a signal that passes the resource UID so once clicked, it loads and applies the changes to the mesh, this also allows me to save the UID in a variable, which I use to set the correct textures when the game is loaded.

When I use the editor everything works fine, using `print(ResourceLoader.get_resource_uid(iris_style))` returns `6360883855146557548` which means it's working as it should. However when I export the project and execute it, every single time I print these IDs (for debugging purposes) it returns `-1` which also means no textures are applied when the thumbnail is selected.

I already tried adding ".remap" to the resource path, but it still returns `-1`. Is this a bug? Or is it supposed to work like that? I can try to find a workaround, but what confuses me is that it works fine in the editor, and it seems like I can't expect the same behavior when I export my project. Isn't this the way I am supposed to handle resource ids? I already tried using `get_rid()` but it also returns 0 or invalid.

### Steps to reproduce

1. Create a few custom resources.
2. Load them into the game and print their IDs using `print(ResourceLoader.get_resource_uid(the_resource))`.
3. Now export the project and check the logs/console outputs (which should be invalid IDs). Make sure it is executed out of the project folder, or it will return the correct IDs.

### Minimal reproduction project

Very small project with 3 custom resources, it returns -1 as expected.

[id_project.zip](https://github.com/godotengine/godot/files/11138864/id_project.zip)

**Comment by KoBeWi** [comment]
Looks like export bug (or maybe not?). The resources are converted to `res` and no longer have UID. You can still load them using `uid://` because it's cached, but you can't get UID from the resource.

**Comment by AThousandShips** [comment]
The UIDs are not carried over in the code, for resources it is done here:
https://github.com/godotengine/godot/blob/ef025711a694071a1407b5035a6fe3e67c2a5b07/editor/export/editor_export_platform.cpp#L795-L803

Could be an oversight, but UIDs mainly make sense in the editor for handling files being moved etc., but still might be good to have it fixed

**Comment by leonardoraele** [comment]
Keeping track of resources' UIDs in exported builds would be useful for multiplayer games. You could send a RPC from one peer to another referencing a resource by its UID instead of its path. (e.g. to instantiate a scene) Since UID is usually shorter than the path, it would be more network efficient than sending the path in the RPC.

**Comment by poohcom1** [comment]
>UIDs mainly make sense in the editor for handling files being moved etc., but still might be good to have it fixed

If this isn't planned to be implemented soon, I think it should be documented clearly in all places referencing resource UIDs, as well as have UID related functions like `get_resource_uid` fail loudly in exported runtime. Many people (like me xD) might end up coding large part of their game with the expectation that UIDs work in runtime because they don't regularly test in exported versions.

**Comment by Summersay415** [comment]
Can confirm.
This can be fixed by looking for .remap-like file and fetching path from them, or something like that. I will try to fix this.

**Comment by KoBeWi** [comment]
Yeah there is `uid_cache.bin` file that you can use for lookup. Other UID sources (like the new `.uid` files) don't/might not exist after export.

**Comment by sergioabreu-g** [comment]
Is there any other way to get the resource UID at runtime? 

I'm making a save system for my game and I can't find another reliable way to store a reference to a project resource. For example, I might want to save an array of pets to a file, where each element is a reference to a `Dog.tres`, `Lion.tres`... I was going to use the UID but I can't due to this. How is this usually done in Godot? I guess there's another reasonable way to do it or this bug would have been found before.

**Comment by poohcom1** [comment]
> Is there any other way to get the resource UID at runtime?

@sergioabreu-g You can just use regular resource paths, like `res://objects/Dog.tres`. UIDs are mapped to resource paths under the hood anyway. However, you lose some refactoring flexibility because moving around files would break references.

If you want to keep using UIDs, one solution I found was to create an [ExportPlugin](https://docs.godotengine.org/en/stable/classes/class_editorexportplugin.html) that would generate a Dictionary of UIDs to paths. Godot resource loading is flexible and works with both `res://` and `uid://`, so you could set up a system that replaces uid with filepaths only in exported builds.

**Comment by sergioabreu-g** [comment]
@KoBeWi oh that's much better than my solution, thank you!

**Comment by KetsuiGit** [comment]
Hello, I would like to report that this issue is still happening on Godot. I am using the version 4.4.1 stable (mono/c#) on MacOS.

You can see that when using the editor with the following script, all the UIDs are working correctly.

<img width="2110" height="1006" alt="Image" src="https://github.com/user-attachments/assets/2e2395fe-874e-4ac7-ab29-ee270a15ab13" />


However, here is the output when I try to export the project and run it on MacOS.

<img width="940" height="512" alt="Image" src="https://github.com/user-attachments/assets/3bead5f8-857d-4caf-9c3c-3846ab7d635e" />

The method ResourceLoader.GetResourceUid(...) returns -1 which is an invalid uid of course.

**Comment by KetsuiGit** [comment]
If anyone come back here, here is the workaround I found. In my setup I had every UID saved somewhere, so I was able to do:

```csharp
    private readonly Dictionary<string, string> _pathToUids = new();

    private void CacheUid(string uid)
    {
        long idAsLong = ResourceUid.TextToId(uid);
        string? pathFromUId = ResourceUid.Singleton.GetIdPath(idAsLong);
        if (pathFromUId == null)
        {
            GD.PrintErr($"Path from UID ${uid} not found is");
            return;
        }

        _pathToUids.Add(pathFromUId, uid);
    }

    // Example of how to use it then
    private override void _Ready()
    {
        foreach (string uid in myListOfUids)
        {
            CacheUid(uid);
        }
    }
```

As you can see, we cannot get the UID from a path when exporting a project, but we can get the path from the UID !
Thanks to that, all I had to do was to iterate over all my UID to fill a dictionary that will contains the mapping between path and UIDs. 😄

**Comment by jitspoe** [comment]
I also got burned trying to use this for a save system.  It feels weird to push the use of UIDs in place of file paths then have them not work in exported builds.  I think the behavior should be consistent between editor and exported builds to avoid headaches like this.  I thought the whole point of the ResourceLoader over using file an dir access directly was that it behaved the same between editor and exported builds.  Could possibly just read from the cache behind the scenes in the exported build?

Also, should `uid_cache[buffer.get_string_from_ascii()] = id` be `uid_cache[buffer.get_string_from_utf8()] = id` in @KoBeWi's example in case the resource path has non-ascii characters?  The UID cache method does seem to work consistently between editor and exported builds, though, so that's a good workaround for now.

Here's what I ended up doing for my save system:
```gdscript
# UID's are not saved in the exported builds, so we need to read the cache
var uid_cache : Dictionary[String, int]

func _ready() -> void:
	var f := FileAccess.open("res://.godot/uid_cache.bin", FileAccess.READ)
	var count := f.get_32()
	for i in count:
		var id := f.get_64()
		var leng = f.get_32()
		var buffer := f.get_buffer(leng)
		uid_cache[buffer.get_string_from_ascii()] = id # Should this be get_string_from_utf8?

func get_uid(entity : Node) -> int:
	return uid_cache[entity.scene_file_path] # UIDs not saved
	#return ResourceLoader.get_resource_uid(entity.scene_file_path) - doesn't work in exported builds

# Whatever entity save logic you use can just call get_uid:
func save_entity_data(entity_to_dict_map : Dictionary, non_spawned_entities : Array):
	var entities := []
	# Save the spawned entities
	for entity in spawned_entities_to_save:
		if (is_instance_valid(entity)):
			var dict := save_save_properties_to_dict(entity)
			dict["uid"] = get_uid(entity)
...
```

**Comment by Deployd1** [comment]
Had the same problem like everyone, made a save system using UID's also using UID's for SFX references and other files in code, because I thought it's way more convenient, it's shorter and I can change the file structure. For anyone who still wants to the benefits but now Migrates from UID to normal paths, look at the plugin "filepath_generator" so you can still change the directory structure of your projects with ease..

This is a real setback for me as I have to replace all UID's in my code with it's path. It was not clearly stated that it doesn't work at runtime. I don't even know why you wouldn't make it work at runtime.. Or atleast make it more CLEAR that it doesn't work.

**Comment by kevnls** [comment]
Oh OK. I must be thinking about my years as a DBA and DB keys. ¯\_(ツ)_/¯

That said, I still don't love the idea of using things like names over a unique ID being managed by a system if possible.
